src/server/pkg/storage/fileset/merge.go

```diff
@@ -341,86 +317,66 @@ func newTagSetMergeReader(frs []*FileReader) *TagSetMergeReader {
 }
 
 // Iterate iterates over the tag merge readers in the merged tagset.
-func (tsmr *TagSetMergeReader) Iterate(f func(*TagMergeReader) error) error {
+func (tsmr *TagSetMergeReader) Iterate(cb func(*TagMergeReader) error) error {
 	return tsmr.pq.iterate(func(ss []stream, _ ...string) error {
-		// Convert generic stream to tag stream.
-		var tss []*tagStream
-		for _, s := range ss {
-			if _, ok := s.(*deleteTagStream); ok {
-				tss = nil
-				continue
+		return tsmr.mergeTag(ss, cb)
+	})
+}
+
+func (tsmr *TagSetMergeReader) mergeTag(ss []stream, cb func(*TagMergeReader) error, deleter ...func(string)) (retErr error) {
+	// Convert generic streams to tag streams.
+	var i int
+	var tss []*tagStream
+	for ; i < len(ss); i++ {
+		if dts, ok := ss[i].(*deleteTagStream); ok {
+			if len(deleter) > 0 {
+				deleter[0](dts.tag.Id)
 			}
-			tss = append(tss, s.(*tagStream))
+			break
 		}
-		if len(tss) == 0 {
-			return nil
+		tss = append(tss, ss[i].(*tagStream))
+	}
+	for i += 1; i < len(ss); i++ {
+		if _, ok := ss[i].(*deleteTagStream); ok {
+			continue
 		}
-		// Check if we are at the padding tag, which indicates we are done.
-		tag, err := tss[0].fr.PeekTag()
-		if err != nil {
+		// TODO: This could be more performant.
+		if err := ss[i].(*tagStream).fr.NextTagReader().Get(ioutil.Discard); err != nil {
 			return err
 		}
-		if tag.Id == paddingTag {
-			for _, ts := range tss {
-				if err := ts.fr.NextTagReader().Get(ioutil.Discard); err != nil {
-					return err
-				}
-			}
-			return nil
-		}
-		// Create tag merge reader and execute callback.
-		var trs []*chunk.TagReader
-		for _, ts := range tss {
-			trs = append(trs, ts.fr.NextTagReader())
-		}
-		return f(newTagMergeReader(trs))
-	})
+	}
+	if len(tss) == 0 {
```

Comment:
Shouldn't we do this check before the loop above?

Reply:
This is operating on the tag streams, not the generic streams. If it was before (and operated on the generic streams), then deletions would not be handled properly since we would never get to the point where the deletion callback is executed and we would not progress the other streams past the deleted tag.